File: src/android/SignicatPlugin.java
package com.signicat.plugin;

import org.apache.cordova.*;
import org.json.JSONArray;
import org.json.JSONException;

import com.connectis.sdk.ConnectisSDK;
import com.connectis.sdk.api.configuration.ConnectisSDKConfiguration;
import com.connectis.sdk.api.authentication.AuthenticationResponse;
import com.connectis.sdk.api.authentication.AuthenticationResponseDelegate;
import com.connectis.sdk.api.authentication.ErrorResponseDelegate;
import com.connectis.sdk.internal.authentication.login.LoginFlow;
import com.connectis.sdk.api.authentication.AccessTokenDelegate;
import com.connectis.sdk.api.authentication.Token;
import org.jetbrains.annotations.NotNull;


public class SignicatPlugin extends CordovaPlugin {

    @Override
    public boolean execute(String action, JSONArray args, CallbackContext callbackContext) throws JSONException {
        if ("loginAppToApp".equals(action)) {
            login(args, callbackContext);
            return true;
        }

        if ("getAccessToken".equals(action)) {
            getAccessToken(callbackContext);
            return true;
        }

        return false;
    }

    private void getAccessToken(final CallbackContext callbackContext) {    

        ConnectisSDK.Companion.useAccessToken(
            cordova.getActivity(),
            new AccessTokenDelegate() {
                @Override
                public void handleAccessToken(@NotNull Token accessToken) {
                    callbackContext.success(accessToken.getValue());
                }

                @Override
                public void onError(@NotNull String exception) {
                    callbackContext.error("Error:getAccessToken: " + exception);
                }
            }
        );

    }


    private void login(JSONArray args, CallbackContext callbackContext) {

        final String issuer;
        final String clientId;
        final String redirectUri;
        final String scopes;
        final String brokerDigidAppAcs;
        final boolean isAppToApp;

        try {
            issuer = args.getString(0);
            clientId = args.getString(1);
            redirectUri = args.getString(2);
            scopes = args.getString(3);
            brokerDigidAppAcs = args.getString(4);
            isAppToApp = args.optBoolean(5, false);

        } catch (JSONException e) {
            callbackContext.error("Invalid args: " + e.getMessage());
            return;
        }

        final boolean allowDeviceAuthentication = false;

        cordova.getActivity().runOnUiThread(() -> {
          try {
            ConnectisSDKConfiguration configuration = new ConnectisSDKConfiguration(
                issuer,
                clientId,
                redirectUri,
                scopes,
                null,
                brokerDigidAppAcs,
                (isAppToApp) ? LoginFlow.APP_TO_APP : LoginFlow.WEB
            );
        
            AuthenticationResponseDelegate delegate = new AuthenticationResponseDelegate() {
              @Override
              public void handleResponse(AuthenticationResponse response) {
                callbackContext.success(response.toString());
              }
        
              @Override
              public void onCancel() {
                callbackContext.error("Signicat login cancelled");
              }
            };
        
            ErrorResponseDelegate errorDelegate = new ErrorResponseDelegate() {
              @Override
              public void handleError(Exception e) {
                callbackContext.error(e.getMessage());
              }
            };
        
            ConnectisSDK.Companion.login(
                configuration,
                cordova.getActivity(),
                delegate,
                errorDelegate,
                allowDeviceAuthentication
            );
          } catch (Exception e) {
            callbackContext.error("Login config error: " + e.getMessage());
          }
        });

    }
}

File: src/ios/SignicatPlugin.swift
import Foundation
import ConnectisSDK
import UIKit



@objc(SignicatPlugin)
class SignicatPlugin: CDVPlugin, AuthenticationResponseDelegate, AccessTokenDelegate  {

    private var currentCommand: CDVInvokedUrlCommand?
    private var accessTokenCallbackId: String?

    @objc(getAccessToken:)
    @MainActor
    func getAccessToken(command: CDVInvokedUrlCommand) {

        self.accessTokenCallbackId = command.callbackId

        showMessage(messageIn: "getAccessToken")

        ConnectisSDK.useAccessToken(
            caller: self.viewController,
            delegate: self
        )

    }

    func handleAccessToken(accessToken: Token) {

        let result = CDVPluginResult(
            status: CDVCommandStatus_OK,
            messageAs: accessToken.getValue()
        )
        self.commandDelegate.send(
            result,
            callbackId: self.accessTokenCallbackId
        )

    }

    func onError(errorMessage: String) {

        let result = CDVPluginResult(
            status: CDVCommandStatus_ERROR,
            messageAs: errorMessage
        )
        self.commandDelegate.send(
            result,
            callbackId: self.accessTokenCallbackId
        )

    }


    @objc(loginAppToApp:)
    @MainActor
    func loginAppToApp(command: CDVInvokedUrlCommand) {

        NSLog("BADRUZ!");

        self.currentCommand = command
/*
        /*Demo app default configuration*/
        let issuer = "https://pkio.broker.ng-test.nl/broker/sp/oidc"
        let clientID = "PRlEjCDjGEzzLimcNOYWnmxY4IWqRHe3"
        let redirectURI = "https://pkio.broker.ng-test.nl/broker/app/redirect/response"
        let appToAppScopes = "openid idp_scoping:https://was-preprod1.digid.nl/saml/idp/metadata_app"
        let brokerDigidAppAcs = "https://pkio.broker.ng-test.nl/broker/authn/digid/digid-app-acs"
        
        /*APalma Signicat SDK - OIDC Client configuration */
        let issuer = "https://preprodbroker.salland.nl/auth/open"
        let clientID = "sandbox-victorious-chess-790"
        let redirectURI = "https://salland-dev.outsystems.app/Adriano_Sandbox/Redirect"
        let appToAppScopes = "openid profile"
        let brokerDigidAppAcs = "https://preprodbroker.salland.nl/broker/authn/digid/acs"

*/
        guard command.arguments.count >= 5,
            let issuer = command.arguments[0] as? String,
            let clientID = command.arguments[1] as? String,
            let redirectURI = command.arguments[2] as? String,
            let appToAppScopes = command.arguments[3] as? String,
            let brokerDigidAppAcs = command.arguments[4] as? String,
            let isAppToApp = command.arguments[5] as? Bool
        else {
            let result = CDVPluginResult(
                status: CDVCommandStatus_ERROR,
                messageAs: "Missing or invalid parameters"
            )
            self.commandDelegate.send(result, callbackId: command.callbackId)
            return
        }




        let configuration = ConnectisSDKConfiguration(
            issuer: issuer,
            clientID: clientID,
            redirectURI: redirectURI,
            scopes: appToAppScopes,
            brokerDigidAppAcs: brokerDigidAppAcs,
            loginFlow: isAppToApp ? LoginFlow.APP_TO_APP : LoginFlow.WEB
        )


        ConnectisSDK.logIn(
            sdkConfiguration: configuration,
            caller: self.viewController,
            delegate: self,
            allowDeviceAuthentication: false//ConnectisSDK.isDeviceAuthenticationEnabled()
        )
    }


    func handleResponse(authenticationResponse: AuthenticationResponse) {

        showMessage(messageIn: "handleResponse")

        guard let command = currentCommand else { return }


        guard let nameId = authenticationResponse.nameIdentifier else {

            if !authenticationResponse.isSuccess {
                let errorMessage = authenticationResponse.error?.localizedDescription ?? "Unknown authentication error"

                let pluginResult = CDVPluginResult(
                    status: CDVCommandStatus_ERROR,
                    messageAs: "Authentication error: \(errorMessage)"
                )

                self.commandDelegate.send(pluginResult, callbackId: command.callbackId)
                self.currentCommand = nil
                return
            }

            let pluginResult = CDVPluginResult(
                status: CDVCommandStatus_ERROR,
                messageAs: "Authentication error"
            )

            self.commandDelegate.send(pluginResult, callbackId: command.callbackId)
            self.currentCommand = nil
            return
        }

        let result: [String: Any] = [
            "nameIdentifier": nameId,
            "isSuccess": authenticationResponse.isSuccess
        ]

        let pluginResult = CDVPluginResult(
            status: CDVCommandStatus_OK,
            messageAs: result
        )

        self.commandDelegate.send(pluginResult, callbackId: command.callbackId)
        self.currentCommand = nil
    }


    func onCancel() {

        showMessage(messageIn: "onCancel")

        guard let command = currentCommand else { return }

        let pluginResult = CDVPluginResult(
            status: CDVCommandStatus_ERROR,
            messageAs: "Authentication was canceled!"
        )

        self.commandDelegate.send(pluginResult, callbackId: command.callbackId)
        self.currentCommand = nil
    }


    func showMessage(messageIn: String){

        let toastController: UIAlertController =
            UIAlertController(
            title: "WOOOOW!",
            message: messageIn,
            preferredStyle: .alert
            )

        toastController.addAction(UIAlertAction(
            title: "OK", 
            style: .default, 
            handler: { _ in 
                print("OK tap") 
            }))

        self.viewController?.present(
            toastController,
            animated: true,
            completion: nil
        )

    }

}

